### PR TITLE
End to end test cases for `st2chatops` bad ST2 config.

### PR DESCRIPTION
**Pull request by jinpingh**
Fixes https://github.com/StackStorm/st2chatops/issues/124
End to end test cases for `st2chatops` invalid ST2 authentication and API URL.
During st2chatops starts up, st2chatops.env is source’d directly and overwrites environment variable value passed from command line unless the it is not enabled.
In order to test different authentication configurations, disable all authentication related variables before test (SETUP) and revert changes after test (TEARDOWN)

### CODE REVIEW

File: chatops/test-hubot-st2-badconfig.bats [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	         ")"
+
+	assert_failure
+	assert_equal $status 1
+	assert_output --partial "Failed to retrieve commands from "$ST2_API_URL""
+}
+
+@test "exit with error status from only providing user name for authentication." {
+	run eval "("\
+	         " cd /opt/stackstorm/chatops; "\
```
Comment (armab) [inline chatops/test-hubot-st2-badconfig.bats]: Can we change current working directory only once to avoid running `cd /opt/stackstorm/chatops` on every test? This will simplify code a bit

File: chatops/test-hubot-st2-badconfig.bats [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+load '../test_helpers/bats-support/load'
+load '../test_helpers/bats-assert/load'
+
+
+ST2_CHATOPS_ENV_FILE='/opt/stackstorm/chatops/st2chatops.env'
+
+@test "SETUP: setup environment variables for authentication related variables." {
+	`sed -e '/ ST2_API_KEY=/s/^/#/g' -i $ST2_CHATOPS_ENV_FILE`
+	`sed -e '/ ST2_AUTH_TOKEN= /s/^/#/g' -i $ST2_CHATOPS_ENV_FILE`
+	`sed -e '/ ST2_AUTH_USERNAME= /s/^/#/g' -i $ST2_CHATOPS_ENV_FILE`
```
Comment (armab) [inline chatops/test-hubot-st2-badconfig.bats]: Can you please help me to understand what's the logic/reason behind this block and why do we need it? I found test description not clear enough

Comment (jinpingh) [reply chatops/test-hubot-st2-badconfig.bats]: @armab During `st2chatops` starts up, `st2chatops.env` is source’d directly and overwrites environment variable value passed from command line unless the it is not enabled.
In order to test different authentication configurations, disable all authentication related variables before test and revert changes after test.

File: chatops/test-hubot-st2-badconfig.bats [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	`sed -e '/ ST2_AUTH_PASSWORD= /s/^/#/g' -i $ST2_CHATOPS_ENV_FILE`
+	`sed -e '/ ST2_AUTH_URL= /s/^/#/g' -i $ST2_CHATOPS_ENV_FILE`;
+}
+
+@test "exit with error status from unresolved ST2_API_URL." {
+	run eval "("\
```
[the remark this answers is not in the document]
Comment (jinpingh) [reply chatops/test-hubot-st2-badconfig.bats]: Have `eval` because check for return status and make sure its value is 1 `assert_equal $status 1`
Maybe there are better way to do it.  Will do more investigations and clean up unnecessary new line.

[the remark this answers is not in the document]
Comment (armab) [reply chatops/test-hubot-st2-badconfig.bats]: Yes, see https://github.com/ztombol/bats-assert
Please try `run` without `eval` and see if that works.

I think it could be addressed by https://github.com/StackStorm/st2tests/pull/170/files#r288962412

File: chatops/test-hubot-st2-badconfig.bats (hunk 2 above)
Comment (armab) [reply chatops/test-hubot-st2-badconfig.bats]: OK, after looking closer at tests you want to `unset` those ENV variables in `st2chatops.env` to verify edge cases.

Please make `@test` comment more clear as `setup environment variables for authentication related variables` is too blurry.

File: chatops/test-hubot-st2-badconfig.bats (hunk 3 above)
[the remark this answers is not in the document]
Comment (jinpingh) [reply chatops/test-hubot-st2-badconfig.bats]: @armab Without `eval`,  it can't handle environment variable passed to `st2chatops`
```✗ exit with error status from unresolved ST2_API_URL.
   (from function `assert_failure' in file st2tests/chatops/../test_helpers/bats-assert/src/assert.bash, line 149,
    in test file st2tests/chatops/test-hubot-st2-badconfig.bats, line 20)
     `assert_failure 1' failed
     /usr/lib/bats/bats-exec-test: line 58: ST2_API_URL=http://non-existent:9101/: No such file or directory

   -- command failed as expected, but status differs --
   expected : 1
   actual   : 127
   output   : /usr/lib/bats/bats-exec-test: line 58: ST2_API_URL=http://non-existent:9101/: No such file or directory
   --```

File: chatops/test-hubot-st2-badconfig.bats [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+load '../test_helpers/bats-support/load'
+load '../test_helpers/bats-assert/load'
+
+
+ST2_CHATOPS_ENV_FILE='/opt/stackstorm/chatops/st2chatops.env'
+cd /opt/stackstorm/chatops
+
+@test "SETUP: make sure ENV variables are unset in st2chatops.env to test edge cases" {
+	`sed -e '/ ST2_API_KEY=/s/^/#/g' -i $ST2_CHATOPS_ENV_FILE`
+	`sed -e '/ ST2_AUTH_TOKEN= /s/^/#/g' -i $ST2_CHATOPS_ENV_FILE`
```
Comment (armab) [inline chatops/test-hubot-st2-badconfig.bats]: Additional request to double check if this all really works.

I'm not 100% sure as I didn't ran it, but whitespace after `ST2_AUTH_TOKEN= ` makes me think there could be issues in similar `sed` rules.

Comment (jinpingh) [reply chatops/test-hubot-st2-badconfig.bats]: @armab You are right about whitespace. Will remove it.

**Comment by armab** [comment]
Also curious, who long does it take to run this set of `chatops/test-hubot-st2-badconfig.bats` tests ?

**Comment by jinpingh** [comment]
It takes 5 seconds around.

**Comment by armab** [comment]
Awesome, that's fast!
Thanks for checking.

**Comment by jinpingh** [comment]
Unit tests covered these test cases, close it without merge.
